packages/teams-js/src/internal/utils.ts

```diff
@@ -54,18 +62,28 @@ export function validateOrigin(messageOrigin: URL): boolean {
     return false;
   }
   const messageOriginHost = messageOrigin.host;
-
-  if (validOrigins.some((pattern) => validateHostAgainstPattern(pattern, messageOriginHost))) {
-    return true;
+  let isValid = Communication.validatedDomains.get(messageOrigin.host);
+  if (isValid === undefined) {
+    validateOriginHandler(messageOrigin).then((value: boolean) => {
+      isValid = value;
+      Communication.validatedDomains.set(messageOrigin.host, value);
+    });
   }
-
-  for (const domainOrPattern of GlobalVars.additionalValidOrigins) {
-    const pattern = domainOrPattern.substring(0, 8) === 'https://' ? domainOrPattern.substring(8) : domainOrPattern;
-    if (validateHostAgainstPattern(pattern, messageOriginHost)) {
+  //If the DDL designates that the URL doesn't exist fallback to fallback list or check if it exists in the supplied validDomains list
+  if (!isValid) {
+    if (validOrigins.some((pattern) => validateHostAgainstPattern(pattern, messageOriginHost))) {
       return true;
     }
-  }
 
+    for (const domainOrPattern of GlobalVars.additionalValidOrigins) {
+      const pattern = domainOrPattern.substring(0, 8) === 'https://' ? domainOrPattern.substring(8) : domainOrPattern;
+      if (validateHostAgainstPattern(pattern, messageOriginHost)) {
+        return true;
+      }
+    }
+  } else {
+    return true;
```

Comment:
Having many nested `if`s can sometimes make code difficult to read, particularly the longer it gets. Two potential approaches to consider:

1. Rather than writing this as "if !isValid { do some longish code} else return true, consider writing this as "if isValid return true; now do some longish code"
The idea being that when people read the code, when they start reading the longish code, they can already know "Okay we've already gotten the heck out of this function if we established validity" and you don't make people wait a long time (and read the code for handling the opposite case) to figure out "ooh what happens if `isValid` was true"